# For more information see: http://www.bioconductor.org/help/workflows/rnaseqGene/
library("DESeq2")
library("tidyverse")
library("apeglm")
library(ggplot2)

##### Set up working directory #####
workdir <- '/Users/jessicawestfall/Dropbox/Mac/Documents/Dowell/Exp83_PROseq/fc_metadata_5ptrunc/'
dir.create(workdir, showWarnings = FALSE)
setwd(workdir)
getwd()
outdir <- paste(workdir, 'deseqresults', '/', sep='') ##naming our outdir
dir.create(outdir, showWarnings = FALSE) ###creating the directory

##### GTF annotation files #####
hg38GTF_5ptrunc <- read.csv('/Users/jessicawestfall/Dropbox/Mac/Documents/Dowell/Exp83_PROseq/hg38_refseq_diff53prime_5ptrunc.gtf', row.names=NULL, sep="\t")
hg38GTF_5ptrunc_noChr21 <- subset(hg38GTF_5ptrunc, chr1!="chr21")

##### Import counts table (featureCounts) and metadata table #####
#read the count table
counts <- read.csv('fc_2reps_5ptrunc/fc_2rep_5ptrunc_counts.csv', row.names = 1, sep="\t")
head(counts) #your rowname should be the gene ids. Your colnames should match some column of your metadata table (in this case filetable)

# read the metadata
metadata <- read.csv("metadata.txt", header=TRUE, sep=",")
head(metadata) 

#Your metatdata columns and your counts rows must be in the same order!!!!!!
countdat <- counts %>% select(as.vector(metadata$label))
head(countdat)

##### DESeq2 on full model #####
# Generate DESeqDataSet from count matrix generated by featureCounts
dds <- DESeqDataSetFromMatrix(countData = countdat, 
                                 colData = metadata, 
                                 design = ~genotype+treatment+genotype:treatment)

### Run DESeq on the DESeqDataSet object
DEdds <- DESeq(dds)

### Get normalized counts for downstream analysis
normalizedcounts = counts(DEdds, normalize=TRUE)
write.csv(normalizedcounts, file = paste0(outdir,"deseq2_normalized_all_PROcounts.txt"))

### output the results for a specified alpha value
alpha_val <- 0.01
#comparison <- c("genotype", "down_syn", "typical")
comparison <- c("treatment",  "IFN", "BSA")
res <- results(DEdds, alpha = alpha_val, contrast = comparison)
summary(res)
head(results(DEdds))
sum(res$padj < alpha_val, na.rm=TRUE) #total number of sig

### Sort summary list by p-value
resSig <- subset(res, res$padj < alpha_val)
resSig <- resSig[ order( resSig$padj),]
head(resSig)
topGene <- rownames(res)[which.min(res$padj)]
plotCounts(dds, gene=topGene, intgroup = c('genotype', 'treatment'))
write.csv(resSig, file = paste0(outdir,"resSig_IFNvsCtrl.csv"))

### Plot distribution
limits <- c(-100, 100)
dispplot <- plotDispEsts(DEdds, main="Dispersion plot Ctrl vs IFN")
dispplot <- plotDispEsts(DEdds, main="Disomy vs Trisomy")

### Subset the significant genes from the res table out of countdat table
sigGenes <- rownames(res)
sigGeneList <- list() #create an empty list
for (sgene in sigGenes){
  srow <- coveragetable %>% filter(grepl(sgene, rownames(coveragetable)))
  sigGeneList[[sgene]] <- srow
}
sigGeneCoveragetable <- do.call("rbind", sigGeneList)


### MA plot
name <- "MA IFN vs Ctrl"
#name <- "MA Disomy vs Trisomy"
limits <- c(-10, 10)
maplot <- plotMA(res, 
                 main=name, 
                 alpha=alpha_val, ylim=limits,
                 colNonSig = "gray60",
                 colSig = "red",
                 colLine = "grey40",
                 )
with(res[topGene, ], {
  points(baseMean, log2FoldChange, col="dodgerblue", cex=2, lwd=2)
  text(baseMean, log2FoldChange, topGene, pos=2, col="dodgerblue")
})

### Volcano plot
alpha_val <- 0.01 # Threshold on the adjusted p-value
cols <- densCols(res$log2FoldChange, -log10(res$padj))
plot(res$log2FoldChange, 
     -log10(res$padj), 
     col=cols, panel.first=grid(),
     #main="Disomy vs. Trisomy", 
     main="IFN vs Control Volcano plot",
     xlab="Effect size: log2(fold-change)", 
     ylab="-log10(adjusted p-value)",
     pch=20, cex=0.1)
abline(v=0)
abline(v=c(-2,2), col="gray")
abline(h=-log10(alpha_val), col="red")
with(subset(res , padj < alpha_val), points(res$log2FoldChange, -log10(res$padj), pch=20, cex=0.1, col="black"))
up <- rownames(res)[which(res$log2FoldChange > 0 & res$padj < alpha_val )]
with(subset(res[which(rownames(res) %in% up),]), points(log2FoldChange, -log10(padj), pch=20, cex=0.1, col="green"))
down <- rownames(res)[which(res$log2FoldChange< (-0) & res$padj < alpha_val )]
with(res[which(rownames(res) %in% down),], points(log2FoldChange, -log10(padj), pch=20, cex=0.1, col="red"))
# add text
gn.selected <- abs(res$log2FoldChange) > 2 & res$padj < alpha_val 
text(res$log2FoldChange[gn.selected],
     -log10(res$padj)[gn.selected],
     lab=rownames(res)[gn.selected ], cex=0.4)

### Volcano plot with subset of human IFN alpha and human IFN gamma
# res table of significant genes in alpha list
alpha_gene_list <- scan("/Users/jessicawestfall/Dropbox/Mac/Documents/Dowell/Exp96_IFN/gsea_hallmark/HALLMARK_INTERFERON_ALPHA_RESPONSE.v2022.1.Hs.grp", what="", sep="\n")
alpha_gene_list <- as.vector(alpha_gene_list) 
alpha_res <- subset(res, subset=(rownames(res) %in% alpha_gene_list))
alpha_res_sig <- subset(alpha_res, alpha_res$padj < alpha_val )
alpha_res_sig <- alpha_res_sig[ order( alpha_res_sig$log2FoldChange),]
#write.csv(alpha_res_sig, file = paste0(outdir,"alpha_res_sig_IFNvsCtrl.csv"))

# res table of significant genes in gamma list
gamma_gene_list <- scan("/Users/jessicawestfall/Dropbox/Mac/Documents/Dowell/Exp96_IFN/gsea_hallmark/HALLMARK_INTERFERON_GAMMA_RESPONSE.v2022.1.Hs.grp", what="", sep="\n")
gamma_gene_list <- as.vector(gamma_gene_list) 
gamma_res <- subset(res, subset=(rownames(res) %in% gamma_gene_list))
gamma_res_sig <- subset(gamma_res, gamma_res$padj < alpha_val)
gamma_res_sig <- gamma_res_sig[ order( gamma_res_sig$log2FoldChange),]
#write.csv(gamma_res_sig, file = paste0(outdir,"gamma_res_sig_IFNvsCtrl.csv"))

with(res, plot(log2FoldChange, -log10(padj), pch=16, cex=0.6, main="Volcano plot with alpha and gamma overlay", xlim=c(-5,5))) 
abline(v=0)
abline(h=-log10(alpha_val), col="grey40")
with(subset(alpha_res_sig, subset=(rownames(alpha_res_sig) %in% alpha_gene_list)), points(log2FoldChange, -log10(padj), pch=16, cex=0.6, col="orange"))
with(subset(gamma_res_sig, subset=(rownames(gamma_res_sig) %in% gamma_gene_list)), points(log2FoldChange, -log10(padj), pch=16, cex=0.6, col="light blue"))

#######################################################
### Subset datatable into control(BSA) only samples ###
######################################################
bsa_counts <- counts[,c(grep("_BSA",names(counts))), drop=FALSE]
bsa_meta <- filetable[ which(filetable$treatment=='BSA'),]

# Generate DESeqDataSet from count matrix generated by featureCounts
ddsBSA <- DESeqDataSetFromMatrix(countData = bsa_counts, colData = bsa_meta, design = ~genotype)
DEdds_BSA <- DESeq(ddsBSA)

normalizedcountsBSA = counts(DEdds_BSA, normalize=TRUE)
write.csv(normalizedcountsBSA, file = paste0(outdir,"deseq_normalized_BSAonly_PROcounts.txt"))

### output the results for a specified alpha value
alpha_val <- 0.01
comparison <- c("genotype", "down_syn", "typical")
resBSA <- results(DEdds_BSA, alpha = alpha_val, contrast = comparison)
summary(resBSA)
sum(resBSA$padj < alpha_val, na.rm=TRUE)

resSigBSA <- subset(resBSA, resBSA$padj < alpha_val )
resSigBSA<-resSigBSA[ order( resSigBSA$padj ),]
#write.csv(resSigBSA, file = paste0(outdir,"resSig_D21vsT21_BSAtreatedonly.csv"))

### MA plot
limits <- c(-10, 10)
maplot <- plotMA(resBSA, 
                 main="Disomy vs Trisomy", 
                 alpha=alpha, ylim=limits,
                 colNonSig = "gray60",
                 colSig = "red",
                 colLine = "grey40",
                 )

### Volcano plot
cols <- densCols(resBSA$log2FoldChange, -log10(resBSA$padj))
plot(resBSA$log2FoldChange, -log10(resBSA$padj), 
     col=cols, panel.first=grid(),
     main="Disomy vs Trisomy", xlab="Effect size: log2(fold-change)", 
     ylab="-log10(adjusted p-value)",
     pch=20, cex=0.1)
abline(v=0)
abline(v=c(-1,1), col="grey40")
abline(h=-log10(alpha), col="grey40")
with(subset(resBSA , padj<0.01), points(resBSA$log2FoldChange, -log10(resBSA$padj), pch=20, cex=0.1, col="black"))
up <- rownames(resBSA)[which(resBSA$log2FoldChange>2 & resBSA$padj < alpha_val)]
with(subset(resBSA[which(rownames(resBSA) %in% up),]), points(log2FoldChange, -log10(padj), pch=20, cex=0.1, col="green"))
down <- rownames(resBSA)[which(resBSA$log2FoldChange<(1*-2) & resBSA$padj < alpha_val)]
with(resBSA[which(rownames(resBSA) %in% down),], points(log2FoldChange, -log10(padj), pch=20, cex=0.1, col="red"))

#######################################################
### Subset datatable into treatment(IFN) only samples ###
######################################################
ifn_counts <- counts[,c(grep("_IFN",names(counts))), drop=FALSE]
ifn_meta <- filetable[ which(filetable$treatment=='IFN'),]

# Generate DESeqDataSet from count matrix generated by featureCounts
ddsIFN <- DESeqDataSetFromMatrix(countData = ifn_counts, colData = ifn_meta, design = ~genotype)
DEdds_IFN <- DESeq(ddsIFN)

normalizedcountsIFN = counts(DEdds_IFN, normalize=TRUE)
write.csv(normalizedcountsIFN, file = paste0(outdir,"deseq_normalized_IFNonly_PROcounts.txt"))
